docs/4.x/archiving-behavior-specification.md

```diff
@@ -0,0 +1,385 @@
+to get the rest of the site information and move on to the next.
+
+**duplicate invalidations are found in the archive_invalidations table**
+
+Duplicate invalidations are disallowed when inserting through ArchiveInvalidator. We specifically look for existing
+invalidations before inserting. However, certain edge cases can allow duplicates to exist. It's also possible for other
+code or users to end up inserting data into those tables.
+
+Duplicates are detected during processing by QueueConsumer. Because of the ordering imposed when we select the next
+archive to execute, duplicate archives will be right next to each other. We'll encounter one right after the other.
+
+Since we look for a batch of archives to run, we'll keep looking for invalidations before starting a set of archive jobs.
+After each query for the next invalidated archive, we'll check in the batch if there's a duplicate, and if so, we skip the
+current one and query again.
+
+Note, there is a currently unhandled edge case: if the last job in a batch has duplicates, at least one of the other
+duplicates will also be run, because we won't query for those invalidations until the first batch finishes. Duplicates
+are an edge case to begin with so we haven't handled this.
+
+**how and when the ttl is checked**
+
+Each period type has an associated TTL that is sometimes used to check if an archive is still valid. These TTLs are only
+used if certain prerequisites are true:
+
+- if the period is not a range and includes today, then it is used. We assume today will get new visits so we make sure
+  archives containing today are always eventually reprocessed. By contrast, for data that is tracked in the past, we
+  explicitly invalidate the period. We don't do this for `today`, because invalidating data for every visit encountered
+  today would make the tracker far less performant. Tracking data in the past, however, does not happen with nearly as
+  high frequency or volume.
+
+- if the period is a range, then we always recompute it when it is no longer valid. This is because ranges are just two
+  arbitrary dates. We cannot preprocess every single permutation during core:archive, so instead we launch the archiving process
+  in browser requests when the ttl expires. Since range archives just aggregate day periods together, and we don't launch
+  archiving for those days, this is still performant.
+
+The TTLs are determined by user and config settings. The day period can be specific in the UI in the System Settings page
+and through INI config. The other periods can have custom TTLs in INI config as well, but not in the UI. See above for
+the specific settings.
+
+If no custom TTL is specified for a non-day period, it defaults to the day TTL.
+
+**amount of concurrency to expect**
+
+In general, we can expect that individual core:archive processes will not process multiple sites at the same time. It is
```

Comment:
I believe technically we can expect multiple sites to be launched at the same time? Even the same site.

Eg if there is only one site, then

* We launch archiver, queue will be eg `$idsites=[1]` and after starting command the queue will be empty
* Next archiver will launch the queue is empty and will be reset to `$idsites=[1]` and the same site will be archived.

Or do we maybe prevent this?

Same could happen from time to time when there are multiple sites... especially if one or few sites take longer to archive while others are faster then the queue will be empty and we reset the list of sites to archive etc.

Or maybe I'm not understanding what is meant by "will not process multiple sites at the same time"

Reply:
actually I see now you're mentioning it re SharedSiteIds

Reply:
It mentions that it's possible w/ the current implementation and probably happens regularly if rarely, but it's not the main use case. And that we want things to work when that happens, but are not too concerned with the performance penalty from the edge case.